Feature branch

Enaah opened the pull request:
### Extending and Forking The "larval-flash-notifications"

When I got in touch with Laravel I watch the [Laravel 5 Fundamentals](https://laracasts.com/series/laravel-5-fundamentals).
In one Episode the "Flash Messaging" is shown and a Laracasts-Helper is used. This Helper was to static for me so I looked for an alternative. 
After a quick websearch [laravel-flash-notifications](https://github.com/skrajewski/laravel-flash-notifications) seemed to do the job.
But there was no option to do a "slideUp"-Message without an X.

> as shown in the [Flash Messaging-Episode](https://laracasts.com/series/laravel-5-fundamentals/episodes/20)

This was essential for me so I modded the GitHub-Project.
#### Examples

> produced in a similar environment as produced in the _Laravel-Fundamentals_
###### Example One

Controller (_ArticlesController.php_):

``` php
    public function store(Requests\ArticleRequest $request) {

      Article::create($request->all());

      Flash::success(TRUE, "Saved Article", "Your Article has been saved.");

      return redirect("articles");
    }
```

VIEW:
<img width="1440" alt="importantview" src="https://cloud.githubusercontent.com/assets/10073766/9112333/467a2bd4-3c4e-11e5-9878-37478708d9b1.png">

> permanent FlashMessage with a X and a Title
###### Example Two

Controller (_ArticlesController.php_):

``` php
    public function store(Requests\ArticleRequest $request) {

      Article::create($request->all());

      Flash::success(TRUE, "Saved Article", "Your Article has been saved.");

      return redirect("articles");
    }
```

VIEW:

<img width="1440" alt="slideupview" src="https://cloud.githubusercontent.com/assets/10073766/9112439/e3df7d5c-3c4e-11e5-822f-c575e80e371e.png">

> slideUp FlashMessage without a X and without a Title

_I hope you going to like it_
I'm looking forward hearing from you!

Regards,
Enaah

Enaah committed:
Adjust the README.md

Adjust the Script-Examples, the structures of the „flash.alerts“-Var
and add a few References

Enaah committed:
Add the boolean-Var 'important' and the possible exceptions

The „flash.alert“-Var has now at least three (max four) Values.
Add a possible exception which could occur when no boolean-Value is
provided.

Enaah committed:
Adjust the Switch-function

The $args-Var contains at least two and at most three Values.
(The $title-Var is still optional)

Enaah committed:
Adjust the "flash.blade.php"-Partial

Extend the Partial by an „if“-Funktion.
Add a short Script at the End which provides a simple „slideUp“
(jQuery is REQUIRED!)

Enaah committed:
fix a Typo

view summery

skrajewski commented:
Hi @Enaah,

thanks for your contribution. Your solution is very interesting but in my opinion this is not correct way to achieve your main target.

In this case plugin should be responsible only for PHP Code and sending flash data to session. The decision what will happen with these data belongs to developer. We provide a simple reusable template for blade by default by don't forget - blade is a part of Laravel.

I know currently *laravel-flash-notifications* doesn't have any way to send additional data to session. May be if this is key to reach the target - plugin should provides data. Behaviour is depend from application.

Kind regards,
Szymon

Enaah commented:
Hi Szymon,

thanks for the Feedback.
So basically you have a Problem with the *<script>-snippet* used in the [flash.blade.php](https://github.com/Enaah/laravel-flash-notifications/commit/f89dc2b7f5800132818276cde73f99e7521fab79)?!

Because in a nutshell I just added another value (the "$important"-Var (boolean)) to the $args-Array.

It gives the user more flexibility. He can decide if he wants a Flash-Message with an X, or not. (as shown in the Laracasts-Episode "Flash Messaging").

I just added the *<script>-snippet* to demonstrate the usage.

Or am I getting your arguments completely wrong? 

Kind regards,
Enaah

skrajewski commented:
Hi,

not exactly... The first question is naming - what *important* means? It's different for different people. In my opinion this argument in this case it's too general. What makes this parameter you called *important*?

Name should be significant, may be it's should be called *auto_close*, *auto_hide*? 

The second problem is associated with semantic versioning and backwards compatibility. You added new argument at the first position of method header, so if someone update a package his code stop working. 

### Plugin responsibility
This plugin should not have a responsibility to add javascripts behaviour, maybe the missing feature is passing the additional paramteres to the alert?

#### Example:

##### Call method from *Controller*
```php
Flash::error('Oh snap!', 'Something went wrong. Please try again for a few seconds.', [
    'auto_close' => true,
    'close_after' => 5000
]);
```

##### Use own prepare alert template
```php
@if(Session::has('flash.alerts'))
    @foreach(Session::get('flash.alerts') as $alert)

        <div class='alert alert-{{ $alert['level'] }} {{ isset($alert['auto_hide']) ?: 'js-auto-hide' : '' }}' {{ isset($alert['close_after']) ?: 'data-close-after="''. $alert['close_after']) ."' : '' }}>
            <button class="close" type="button" data-dismiss="alert" aria-hidden="true">&times;</button>

            @if( ! empty($alert['title']))
                <div><strong>{{ $alert['title'] }}</strong></div>
            @endif

            {{ $alert['message'] }}
        </div>

    @endforeach
@endif
```

##### Add additional behaviour

```javascript
$(document).ready( function(){
    $('.js-auto-hide').each(function(index, element) {
        var delay = this.getAttribute('data-close-after') || 5000;
        $(this).delay(delay).slideUp(300);
    });
});
```

### Summary
This is only concept but I think it's good example how to build real flexibility - send any value to flash message. Flash plugin provide a **basic and common functionality** but how you use it in your application depends on you.

Kind regads,
Szymon

Enaah commented:
Hi Szymon,

ok, I think I understand what the problem is.
I totally agree with you that *this plugin should not have a responsibility to add javascripts behaviour*!

Also I think you're right with this statement *This plugin should not have a responsibility to add javascripts behaviour*.

I think the best name for the *Var* should have been *show_x*, but thats irrelevant because without JS it makes no sense (as you said).

You're concept is brilliant. It didn't came to my mind when I thought about the Problem, but having all *Variables* in the Controller can be very useful (if you want to have **maximum flexibility**).
(And since I'm not very fluent in JS it's a good template to work with ;) )

I think this Pull-Request can be closed.

This was my first commit, so apologies if it wasn't very useful.

I really dig your plugin!

Keep it up and best wishes,
Enaah

skrajewski commented:
Hi @Enaah,

I'm glad I could help you with your first PR :) It was a great exercise for us. Thank you for your support in open source!

Kind regards,
Szymon
